hibernate-core/src/main/java/org/hibernate/cfg/annotations/CollectionBinder.java

```diff
@@ -1109,15 +1109,28 @@ private static String buildOrderByClauseFromHql(String orderByFragment, Persiste
 		if ( orderByFragment != null ) {
 			if ( orderByFragment.length() == 0 ) {
 				//order by id
-				return "id asc";
+				return buildOrderById( associatedClass, " asc" );
 			}
 			else if ( "desc".equals( orderByFragment ) ) {
-				return "id desc";
+				return buildOrderById( associatedClass, " desc" );
 			}
 		}
 		return orderByFragment;
 	}
 
+	private static String buildOrderById(PersistentClass associatedClass, String order) {
+		final StringBuilder sb = new StringBuilder();
+		final Iterator<Selectable> columnIterator = associatedClass.getIdentifier().getColumnIterator();
+		while ( columnIterator.hasNext() ) {
+			final Selectable selectable = columnIterator.next();
+			sb.append( selectable.getText() );
+			sb.append( order );
+			sb.append( ", " );
+		}
+		sb.setLength( sb.length() - 2 );
```

Comment:
I think the below code change would be slightly better
```
while ( columnIterator.hasNext() ) {
    ... ...
    if ( columnIterator.hasNext() ) {
        sb.append(", ");
    }
}
```

Reply:
I like the branchless loop-body approach more though.

Reply:
Got it. I hate to be focused on such low level details and obviously the PR is so great. But the current approach will trigger exception when the while loop is skipped in the first place. If we can ensure at least one column exists, we might well achieve the goal of branchless coding without any hack:
```
sb.append( columnIterator.next() );
while ( columnIterator.hasNext() ) {
    sb.append(", ");
    sb.append( columnIterator.next() );
}
```
Ignore my comment if needed however.